core/src/test/java/org/apache/iceberg/TestDeleteFiles.java

```diff
@@ -349,6 +352,54 @@ public void testDeleteFilesOnIndependentBranches() {
         statuses(Status.EXISTING, Status.DELETED, Status.DELETED));
   }
 
+  @Test
+  public void testDeleteWithCollision() {
+    Schema schema = new Schema(Types.NestedField.of(0, false, "x", Types.StringType.get()));
+    PartitionSpec spec = PartitionSpec.builderFor(schema).identity("x").build();
+    Table collisionTable =
+        TestTables.create(tableDir, "hashcollision", schema, spec, formatVersion);
+
+    PartitionData partitionOne = new PartitionData(spec.partitionType());
+    partitionOne.set(0, "Aa");
+    PartitionData partitionTwo = new PartitionData(spec.partitionType());
+    partitionTwo.set(0, "BB");
+
+    Assert.assertEquals(
+        StructLikeWrapper.forType(spec.partitionType()).set(partitionOne).hashCode(),
+        StructLikeWrapper.forType(spec.partitionType()).set(partitionTwo).hashCode());
+
+    Metrics metrics = new Metrics(1L, null, null, null, null);
+
+    DataFile testFileOne =
+        DataFiles.builder(spec)
+            .withPartition(partitionOne)
+            .withPath("/g1.parquet")
+            .withFileSizeInBytes(100)
+            .withMetrics(metrics)
+            .build();
+    DataFile testFileTwo =
+        DataFiles.builder(spec)
+            .withPartition(partitionTwo)
+            .withMetrics(metrics)
+            .withFileSizeInBytes(100)
+            .withPath("/g2.parquet")
+            .build();
+
+    collisionTable.newFastAppend().appendFile(testFileOne).appendFile(testFileTwo).commit();
+
+    Assert.assertEquals(
+        "We should have two files",
+        2,
+        Lists.newArrayList(collisionTable.newScan().planFiles().iterator()).size());
+
+    collisionTable.newDelete().deleteFromRowFilter(Expressions.equal("x", "BB")).commit();
+
+    Assert.assertEquals(
+        "We should have deleted one of them",
+        1,
+        Lists.newArrayList(collisionTable.newScan().planFiles().iterator()).size());
```

Comment:
nit: might be a bit redundant, but I am curious if we shall verify the remaining files is correct and we deleted the desired partition instead?

I am thinking about maybe something like 
```java
Assert.assertEquals(
  partitionOne,
  collisionTable.newScan().planFiles().iterator().next().file().partition());
```

Reply:
Sure, that's a great idea.